### DISCUSSION TITLE
Regarding the new building type sorting

### DISCUSSION THREAD

**Discussion by pkoby** [streetcomplete/StreetComplete#3493] [Feedback]
I tried out the latest version with the new method of sorting building type ( #3373 ). I like the idea of this change, but in practice I will admit it was confusing (coming from the expectations and experience with the old configuration, especially).

Yesterday, I went to an area with a lot of different building types (many industrial and shops). When I got to an area with just rowhouses (by far the most common type around here) I had to reselect it from the full menu (not a problem), but then on the second building, the recent selection of =terrace was not in the top three (which is all I can see on my phone without scrolling) and kept changing places as I kept mapping.

If the last selection was always at the top, then when you keep selecting the same type, it's always in the same spot (whereas otherwise it might change places as it gets more "popular"). When I'm mapping terraces, there's always going to be a bunch in a row.

So essentially, my preference would be latest in spot # 1 always, then # 2-6 sorted by popularity. Thus if there are a lot of the same type (even if it's a new type), you'll find it at the top, but if you select one of # 2-6, the previous # 1 drops to its spot in recent popularity.

Thoughts?

**Comment by smichel17** [comment]
First— I'm totally open to changing the behavior, including a full revert, if needed. Unfortunately, the weather in my area got cold (after the most amazingly long and delightful fall in a decade) just as I finished implementing it, so my mapping has slowed quite a bit and I didn't have the chance to field test it very much.

I can also see how it would work less well in an area with many varied building types, whereas in my area most buildings are the same 4 or 5 all the time (so I want those always present, and just the last few changing based on which of the less common types is present in the neighborhood I'm mapping).

> If the last selection was always at the top, then when you keep selecting the same type, it's always in the same spot (whereas otherwise it might change places as it gets more "popular").

This would be easy to implement.

Here's some other ideas I have for possible tweaks. Let me know if any of them sound useful:

1. Show the same entries as current, but sort them by most recently-used.
    - Upside: The order will stay totally stable.
    - Downside: you won't be able to predict which entry will disappear.
2. Always show the most recent 2 entries, instead of just the most recent one. The rest still by "popularity".
3. Keep a longer history (50-100 entries, instead of 30), so the order is less volatile.
4. Keep a much longer history (200+ entries), but clear the weights when the app is closed (ie, keep one entry for each of the current most-popular, so the entries stay the same the next time you open the app, but will start rapidly changing as you map).
    - All the others are trivial to implement; this one is a bit more effort.

**Reply by smichel17** [reply]
When this completes, you can install the debug apk from there to test out your suggestion https://github.com/smichel17/StreetComplete/actions/runs/1431817678

**Comment by pkoby** [comment]
I wouldn't revert it, it's still a good change.

As to your other suggestions:

1. This one would be fine in my opinion. This is pretty much how it used to be, but taking the last 30 into account, right? I think you could predict which would disappear, because it would be at the bottom of the list of six, right?
2. Also fine.
3. In this case, if you start mapping a new type of building, it would stay stuck at position 6 for even longer, right? Essentially it would take longer to climb the list. So this would still annoy me because it involves scrolling down to the bottom few options for quite awhile.
4. I don't think this option would be good, because it would be quite volatile quite frequently.

I'll test the debug today if I get the chance.

**Reply by smichel17** [reply]
4. This would be the least volatile, since the weights would never reset (except when you close and re-open the app). The only way the order would ever change is that whichever entry you selected most recently might climb up a spot on the list. Based on your dislike of (3), I think you probably wouldn't like this.

**Reply by pkoby** [reply]
I close my apps all the time (I've heard this is not RAM-savvy, but whatever). So every time I go surveying it would reset for me. I don't know what most people do, though.

But you're right, either I would dislike it for the same reason as (3) or for the opposite reason. :smile:

**Comment by smichel17** [comment]
> This is pretty much how it used to be, but taking the last 30 into account, right?

Say my recent answers are (earlier is more recent): `B`, `C`, `D`, `E`, `F`, `G`, `G`.
And my next answer is going to be `A`. Before and after answering `A`…

- **Previous** behavior: `G` is dropped because it's the oldest
    - Before: `B`, `C`, `D`, `E`, `F`, `G`
    - After: `A`, `B`, `C`, `D`, `E`, `F`
- **Current** behavior: `F` is dropped because it's (the oldest of) the least-picked. Note: `F` is last.
    - Before: `G`, `B`, `C`, `D`, `E`, `F`
    - After: `G`, `A`, `B`, `C`, `D`, `E`
- **Proposed (1)** behavior: `F` is dropped because it's (the oldest of) the least-picked. Note: `F` is not last.
    - Before: `B`, `C`, `D`, `E`, `F`, `G`
    - After: `A`, `B`, `C`, `D`, `E`, `G`

**Comment by smichel17** [comment]
Okay, I've finally had a few more chances to go mapping with the new sorting. My impressions:

- **When it works, it's like magic.** I really mean that— there were 2 or 3 times when I thought I was going to have to open the full list because some different type of building popped up where I didn't expect it, but then the right type was just there in the suggestions. I was truly astonished because I was so sure there's *no way* it would be there, and then it was… and I'm the one who *wrote* the algorithm!
    - It's hard to say for sure, but based on this, I think 30 entries is a pretty good amount of history to count.
- **The changing order is definitely confusing at times.** There were also 2-3 times when items shifted around on me in confusing ways, that I couldn't predict… and again, if anybody _should_ be able to predict how it works, it's me.

I'm going to start with option 1 (same items, revert to chronological order) from https://github.com/streetcomplete/StreetComplete/discussions/3493#discussioncomment-1602090. I suspect that option 2 (always show the _two_ most recent entries) is also a good idea, but I'm going to hold off on making that change since I did not actually encounter the situation where a recent entry fell off the list just before I wanted to use it again. I'll continue to keep an eye out for it, though (and the chronological order change will make it more obvious when it happens).

**Reply by smichel17** [reply]
https://github.com/streetcomplete/StreetComplete/pull/3564

**Comment by smichel17** [comment]
See https://github.com/streetcomplete/StreetComplete/issues/3593 for more discussion about this, in the context of the building *levels* quest.
